File: impl/ts/src/evaluator.js
const { compile } = require('./visitors/instructions')

class PredicatorEvaluator {
  constructor (instructions, context = {}) {
    this.instructions = instructions
    this.context = context
    this.stack = []
    this.instructionPointer = 0
  }

  push (val) {
    this.stack.push(val)
  }

  pop () {
    return this.stack.pop()
  }

  result () {
    while (this.instructionPointer < this.instructions.length) {
      this.process(this.instructions[this.instructionPointer])
      this.instructionPointer += 1
    }

    return this.pop()
  }

  process (instruction) {
    const op = instruction[0]
    const val = instruction[1]

    switch (op) {
      case 'not': this.push(!this.pop()); break
      case 'array':
      case 'lit': this.push(val); break
      case 'load': this.push(this.context[val]); break
      case 'jfalse': this.jumpIf(false, val); break
      case 'jtrue': this.jumpIf(true, val); break
      case 'compare': this[`compare_${val}`](); break
      case 'to_bool': this.push(!!this.pop()); break
      case 'to_int': this.push(parseInt(this.pop())); break
      case 'to_date': this.push(this.toDate(this.pop())); break
      case 'to_str': this.push(this.toString(this.pop())); break
      case 'date_ago': this.dateAgo(); break
      case 'date_from_now': this.dateFromNow(); break
      case 'blank': this.push(this.isBlank(this.pop())); break
      case 'present': this.push(!this.isBlank(this.pop())); break
    }
  }

  isBlank (val) {
    return (val === undefined || val === "")
  }

  toString (val) {
    return val
  }

  toDate (val) {
    return (new Date(val)).getTime()
  }

  dateAgo () {
    const seconds = parseInt(this.pop())
    const pastTimestamp = Date.now() - (seconds * 1000)
    this.push(this.toDate(pastTimestamp))
  }

  dateFromNow () {
    const seconds = parseInt(this.pop())
    const futureTimestamp = Date.now() + (seconds * 1000)
    this.push(this.toDate(futureTimestamp))
  }

  jumpIf (bool, offset) {
    if (this.stack[this.stack.length - 1] === bool) {
      this.instructionPointer += (offset - 1)
    } else {
      this.pop()
    }
  }

  compare (comparison) {
    const right = this.pop()
    const left = this.pop()

    if (left === undefined || right === undefined) {
      this.push(false)
    } else {
      this.push(comparison(left, right))
    }
  }

  compare_EQ () {
    return this.compare((left, right) => left === right)
  }

  compare_NEQ () {
    return this.compare((left, right) => left !== right)
  }

  compare_GT () {
    return this.compare((left, right) => left > right)
  }

  compare_GTE () {
    return this.compare((left, right) => left >= right)
  }

  compare_LT () {
    return this.compare((left, right) => left < right)
  }

  compare_LTE () {
    return this.compare((left, right) => left <= right)
  }

  compare_IN () {
    return this.compare((left, right) => right.indexOf(left) !== -1)
  }

  compare_NOTIN () {
    return this.compare((left, right) => right.indexOf(left) === -1)
  }

  compare_BETWEEN () {
    const max = this.pop()
    const min = this.pop()
    const val = this.pop()

    if (max === undefined || min === undefined || val === undefined) {
      this.push(false)
    } else {
      const result = (val > min) && (val < max)
      this.push(result)
    }
  }

}

function evaluateInstructions (instructions, context = {}) {
  const evaluator = new PredicatorEvaluator(instructions, context)
  return evaluator.result()
}

function evaluate (text, context = {}) {
  const instructions = compile(text)
  return evaluateInstructions(instructions, context)
}

module.exports = {
  evaluate,
  evaluateInstructions,
  PredicatorEvaluator
}

File: impl/go/evaluator.go
package predicator

import (
	"errors"
	"fmt"
	"strconv"
	"time"
)

func NewEvaluator(instructions [][]interface{}, data map[string]interface{}) *Evaluator {
	e := Evaluator{
		instructions: instructions,
		data:         data,
		stack:        &stack{},
	}
	return &e
}

type Evaluator struct {
	instructions [][]interface{}
	data         map[string]interface{}
	stack        *stack
	ip           int
}

type stack struct {
	nodes []interface{}
	count int
}

func (s *stack) pop() interface{} {
	if s.count == 0 {
		return nil
	}
	s.count--
	return s.nodes[s.count]
}

func (s *stack) push(v interface{}) {
	s.nodes = append(s.nodes[:s.count], v)
	s.count++
}

func (s *stack) peek() interface{} {
	if s.count == 0 {
		return nil
	}
	return s.nodes[s.count-1]
}

const (
	InstructionNot         = "not"
	InstructionJumpIfFalse = "jfalse"
	InstructionJumpIfTrue  = "jtrue"
	InstructionLiteral     = "lit"
	InstructionArray       = "array"
	InstructionLoad        = "load"
	InstructionToBool      = "to_bool"
	InstructionToInt       = "to_int"
	InstructionToString    = "to_str"
	InstructionToDate      = "to_date"
	InstructionDateAgo     = "date_ago"
	InstructionDateFromNow = "date_from_now"
	InstructionBlank       = "blank"
	InstructionPresent     = "present"
	InstructionCompare     = "compare"

	KeywordCompareBetween     = "BETWEEN"
	KeywordCompareEqual       = "EQ"
	KeywordCompareInclude     = "IN"
	KeywordCompareGreaterThan = "GT"
	KeywordCompareLessThan    = "LT"
)

var (
	ErrInvalidInstruction = errors.New("invalid instruction")
)

func (e *Evaluator) result() (output bool) {
	defer func() {
		r := recover()
		if r != nil {
			fmt.Println("recovered: ", r)
			output = false
		}
	}()
	for e.ip < len(e.instructions) {
		e.process(e.instructions[e.ip])
		e.ip = e.ip + 1
	}
	return e.stack.pop().(bool)
}

func (e *Evaluator) process(instruction []interface{}) error {
	i, ok := instruction[0].(string)
	if !ok {
		return ErrInvalidInstruction
	}
	switch i {
	case InstructionNot:
		v, ok := e.stack.pop().(bool)
		if !ok {
			return ErrInvalidInstruction
		}
		e.stack.push(!v)
	case InstructionJumpIfFalse:
		offset, ok := instruction[len(instruction)-1].(int)
		if !ok {
			return ErrInvalidInstruction
		}
		e.jumpIfFalse(offset)
	case InstructionJumpIfTrue:
		offset, ok := instruction[len(instruction)-1].(int)
		if !ok {
			return ErrInvalidInstruction
		}
		e.jumpIfTrue(offset)
	case InstructionLiteral, InstructionArray:
		e.stack.push(instruction[len(instruction)-1])
	case InstructionLoad:
		key, ok := instruction[len(instruction)-1].(string)
		if !ok {
			return ErrInvalidInstruction
		}
		v, ok := e.data[key]
		if !ok {
			return ErrInvalidInstruction
		}
		e.stack.push(v)
	case InstructionToBool:
	case InstructionToInt:
		n, err := e.toInt(e.stack.pop())
		if err != nil {
			return ErrInvalidInstruction
		}
		e.stack.push(n)
	case InstructionToString:
	case InstructionToDate:
		t, err := e.toDate(e.stack.pop())
		if err != nil {
			return ErrInvalidInstruction
		}
		e.stack.push(t)
	case InstructionDateAgo:
		n, ok := e.stack.pop().(int)
		if !ok {
			return ErrInvalidInstruction
		}
		e.stack.push(e.dateAgo(n))
	case InstructionDateFromNow:
		n, ok := e.stack.pop().(int)
		if !ok {
			return ErrInvalidInstruction
		}
		e.stack.push(e.dateFromNow(n))
	case InstructionBlank:
	case InstructionPresent:
	case InstructionCompare:
		next := instruction[len(instruction)-1]
		if next == KeywordCompareBetween {
			// compare_between
		} else {
			v, ok := next.(string)
			if !ok {
				return ErrInvalidInstruction
			}
			e.compare(v)
		}
	}
	return nil
}

func (e *Evaluator) jumpIfFalse(offset int) {
	b, ok := e.stack.peek().(bool)
	if ok {
		if !b {
			e.ip += offset - 1
			return
		}
	}
	e.stack.pop()
}

func (e *Evaluator) jumpIfTrue(offset int) {
	b, ok := e.stack.peek().(bool)
	if ok {
		if b {
			e.ip += offset - 1
			return
		}
	}
	e.stack.pop()
}

func (e *Evaluator) compare(comparison string) {
	right := e.stack.pop()
	left := e.stack.pop()
	if left == nil || right == nil {
		e.stack.push(false)
	} else {
		switch comparison {
		case KeywordCompareEqual:
			e.stack.push(e.compareEquality(left, right))
		case KeywordCompareInclude:
			e.stack.push(e.compareInclude(left, right))
		case KeywordCompareGreaterThan:
			e.stack.push(e.compareGreaterThan(left, right))
		case KeywordCompareLessThan:
			e.stack.push(e.compareLessThan(left, right))
		default:
			e.stack.push(false)
		}
	}
}

func (e *Evaluator) compareEquality(left, right interface{}) bool {
	leftTime, ok := left.(time.Time)
	if ok {
		rightParsed, ok := right.(string)
		if ok {
			rightTime, err := time.Parse("2006-01-02", rightParsed)
			if err != nil {
				return false
			}
			return leftTime.Equal(rightTime)
		}
	}
	return left == right
}

func (e *Evaluator) compareInclude(left, right interface{}) bool {
	xs, ok := right.([]interface{})
	if !ok {
		return false
	}
	for _, x := range xs {
		if left == x {
			return true
		}
	}
	return false
}

func (e *Evaluator) compareGreaterThan(left, right interface{}) bool {
	// handle dates
	leftTime, ok := left.(time.Time)
	if ok {
		rightTime, ok := right.(time.Time)
		if ok {
			return leftTime.After(rightTime)
		}
	}
	return false
}

func (e *Evaluator) compareLessThan(left, right interface{}) bool {
	// handle dates
	leftTime, ok := left.(time.Time)
	if ok {
		rightTime, ok := right.(time.Time)
		if ok {
			return leftTime.Before(rightTime)
		}
	}
	return false
}

func (e *Evaluator) toInt(v interface{}) (int, error) {
	n, ok := v.(int)
	if ok {
		return n, nil
	}
	str, ok := v.(string)
	if ok {
		return strconv.Atoi(str)
	}
	return 0, ErrInvalidInstruction
}

func (e *Evaluator) toDate(v interface{}) (time.Time, error) {
	t, ok := v.(time.Time)
	if ok {
		return t, nil
	}
	str, ok := v.(string)
	if ok {
		return time.Parse("2006-01-02", str)
	}
	return time.Time{}, ErrInvalidInstruction
}

func (e *Evaluator) dateAgo(n int) time.Time {
	seconds := time.Duration(n << 9)
	return time.Now().Add(-seconds).UTC()
}

func (e *Evaluator) dateFromNow(n int) time.Time {
	seconds := time.Duration(n << 9)
	return time.Now().Add(seconds).UTC()
}

// def date_ago seconds
// past_time = Time.now - seconds
// to_date past_time.strftime "%Y-%m-%d"
// end
